File: python/relay_test.py
# Tests that correspond to the `src/relay.rs` rust api.
from iroh import RelayMap, RelayMode, RelayConfig


def test_relay_map_crud():
    m = RelayMap.empty()
    assert m.is_empty()
    assert m.len() == 0

    cfg = RelayConfig(
        url="https://relay.example.org/",
        quic_port=7842,
        auth_token="hunter2",
    )
    m.insert(cfg)
    assert m.len() == 1
    assert m.contains("https://relay.example.org/")

    got = m.get("https://relay.example.org/")
    assert got is not None
    assert got.url == "https://relay.example.org/"
    assert got.quic_port == 7842
    assert got.auth_token == "hunter2"

    assert "https://relay.example.org/" in m.urls()

    assert m.remove("https://relay.example.org/")
    assert m.is_empty()


def test_relay_map_from_urls():
    m = RelayMap.from_urls(["https://r1.example.org/", "https://r2.example.org/"])
    assert m.len() == 2


def test_relay_mode_constructors():
    RelayMode.disabled()
    RelayMode.default_mode()
    RelayMode.staging()
    m = RelayMap.from_urls(["https://r1.example.org/"])
    custom = RelayMode.custom(m)
    assert custom.relay_map().len() == 1
    RelayMode.custom_from_urls(["https://r2.example.org/"])

File: kotlin/lib/src/test/kotlin/computer/iroh/RelayTest.kt
package computer.iroh

import kotlin.test.Test

class RelayTest {
    @Test fun relayMapCrud() {
        val m = RelayMap.empty()
        assert(m.isEmpty())
        assert(m.len() == 0u)

        val cfg = RelayConfig(
            url = "https://relay.example.org/",
            quicPort = 7842u,
            authToken = "hunter2",
        )
        m.insert(cfg)
        assert(m.len() == 1u)
        assert(m.contains("https://relay.example.org/"))

        val got = m.get("https://relay.example.org/")
        assert(got != null)
        assert(got!!.url == "https://relay.example.org/")
        assert(got.quicPort == 7842u.toUShort())
        assert(got.authToken == "hunter2")

        assert(m.urls().contains("https://relay.example.org/"))
        assert(m.remove("https://relay.example.org/"))
        assert(m.isEmpty())
    }

    @Test fun relayMapFromUrls() {
        val m = RelayMap.fromUrls(listOf("https://r1.example.org/", "https://r2.example.org/"))
        assert(m.len() == 2u)
    }

    @Test fun relayModeConstructors() {
        RelayMode.disabled()
        RelayMode.defaultMode()
        RelayMode.staging()
        val m = RelayMap.fromUrls(listOf("https://r1.example.org/"))
        val custom = RelayMode.custom(m)
        assert(custom.relayMap().len() == 1u)
        RelayMode.customFromUrls(listOf("https://r2.example.org/"))
    }
}
